src/contrib/mlir/backend/pass/memory_optimization.cpp

```diff
@@ -43,114 +56,567 @@ namespace
     using namespace ngraph::runtime::ngmlir;
     using namespace mlir;
 
+    // A helper data-structure to track cannot alias relationship between tensor syms
+    // If NoAlias[T] contains S, then T and S cannot alias.
+    // The relationship is transitive and we compute transitive closure on each update to the
+    // the strucutre.
+    class AliasRelation
+    {
+    public:
+        AliasRelation();
+        /// Checks if values a and b can alias
+        bool canAlias(Value* a, Value* b);
+        void insertNoAlias(Value* a, Value* b);
+
+    private:
+        void computeTransitiveClosure();
+#ifdef NGRAPH_DEBUG_ENABLE
+        void checkInvariance();
+#endif
+    private:
+        using Row = SmallVector<int8_t, 10>;
+        std::unordered_map<Value*, unsigned> m_valueToIdx;
+        SmallVector<Row, 10> m_reachability;
+        unsigned m_maxIdx;
+        bool m_needsTransitiveClosure;
+    };
+
+    class Liveness
```

Comment:
Is this an analysis? -> `LivenessAnalysis`?

Reply:
Will rename.